### PR TITLE
adding support for `RANK` and `DENSE_RANK`

### PR DESCRIPTION
**Pull request by JCOR11599**
Fix for:
 - https://github.com/dolthub/dolt/issues/4126

Note: queries in MySQL seem to be ordered by the ranks by default.

### CODE REVIEW

File: sql/expression/function/aggregation/window_functions.go [hunk 1]
```diff
@@ -1211,6 +1211,150 @@ func (a *PercentRank) Compute(ctx *sql.Context, interval sql.WindowInterval, buf
 	}
 }
 
+type Rank struct {
```
Comment (max-hoffman) [inline sql/expression/function/aggregation/window_functions.go:1214]: i'd collapse the 3 rank functions into a `rankBase` if possible, percent rank doesn't need extra fields but you could use an enum field to differentiate the 3, dense rank will require the two extra vars. The compute function will switch on the rank type. Alternatively, if the compute function for the 3 combined becomes really gnarly, you could instead have `Rank`, `PercentRank`, and `DenseRank` overload `Compute` methods, with `DenseRank` and `PercentRank` tracking their own corresponding states relative to the child `baseRank`.

File: enginetest/enginetests.go [hunk 2]
```diff
@@ -3648,6 +3684,24 @@ func TestWindowFunctions(t *testing.T, harness Harness) {
 		{5, 1.0},
 	}, nil, nil)
 
+	TestQueryWithContext(t, ctx, e, harness, `SELECT a, rank() over (partition by c order by b) FROM t1 order by a`, []sql.Row{
```
Comment (max-hoffman) [inline enginetest/enginetests.go:3687]: Windows have a lot of edge cases, need to test different combinations of partition, order, framing for each.

File: sql/expression/function/aggregation/window/dense_rank.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+package window
+
+import (
+	"strings"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/expression/function/aggregation"
+)
+
+type DenseRank struct {
```
Comment (max-hoffman) [inline sql/expression/function/aggregation/window/dense_rank.go:24]: not pressing but the window expressions are starting to look real similar, codegen potential

File: sql/expression/function/aggregation/window_functions.go [hunk 4]
```diff
@@ -1156,59 +1156,136 @@ type PercentRank struct {
...
+
+// Compute returns the number of unique elements before the current peer group (rank) + 1.
+// ex: [1, 2, 2, 2, 3, 3, 3, 4, 5, 5, 6] => every 3 returns uint64(3) because
+// there are 2 unique values less than 3
+func (a *DenseRank) Compute(ctx *sql.Context, interval sql.WindowInterval, buf sql.WindowBuffer) interface{} {
+	rank := a.rankBase.Compute(ctx, interval, buf)
+	if rank == nil {
+		return nil
+	}
+	if a.partitionEnd-a.partitionStart == 1 {
```
Comment (max-hoffman) [inline sql/expression/function/aggregation/window_functions.go:1275]: `rank.(uint64) == 1` already covers this?

**Review by max-hoffman — APPROVED** [review]
lgtm, one extraneous if clause
